### ISSUE TITLE
Regression / breaking behavior change between 26.4.0 and 26.6.0: CI now lints many more files and fails with historical violations

### ISSUE THREAD

**Issue by gforster** [ansible/ansible-lint#5094]
##### Summary

After an unpinned CI install floated from `ansible-lint==26.4.0` to `ansible-lint==26.6.0`, our previously passing GitLab CI lint job began failing.

The same lint command previously passed under `26.4.0`, processing only 4 files, but fails under `26.6.0`, processing significantly more repository content and reporting 242 fatal violations.

This appears to be a breaking behavioral change between `26.4.0` and `26.6.0`, likely related to file discovery, lint target selection, ignore/pathspec handling, or fatal rule handling.

##### Issue Type

* Bug Report

##### OS / ENVIRONMENT

```console
$ ansible-lint --version
[WARNING]: Ansible is being run in a world writable directory (/builds/unix-team/automated-build), ignoring it as an ansible.cfg source. For more information see https://docs.ansible.com/ansible/devel/reference_appendices/config.html#cfg-in-world-writable-dir
ansible-lint 26.6.0 using ansible-core:2.19.11 ansible-compat:26.6.0 ruamel-yaml:0.19.1 ruamel-yaml-clib:0.2.15
```

CI environment:

```console
GitLab Runner: 16.3.0
Executor: Kubernetes
Container image: registry.access.redhat.com/ubi9/python-311:latest
Python: 3.11
```

Previous working versions:

```console
ansible-lint==26.4.0
ansible-core==2.19.11
ansible-compat==26.3.0
yamllint==1.38.0
pathspec==1.0.4
wcmatch==10.1
```

Failing versions:

```console
ansible-lint==26.6.0
ansible-core==2.19.11
ansible-compat==26.6.0
yamllint==1.38.0
pathspec==1.1.1
wcmatch==10.2
```

* ansible installation method: pip
* ansible-lint installation method: pip

##### STEPS TO REPRODUCE

This is from a GitLab CI job running in a Kubernetes executor.

The CI job installs the latest available `ansible-lint` from pip and then runs lint against the repository root:

```console
pip3 install ansible-lint
pip3 install yamllint
ansible-lint --version
ansible-lint --offline --profile=basic -p .
```

The repository also contains a custom `.yamllint` file:

```yaml
---
extends: default

ignore:
  - collections/
  - roles/
  - '!collections/requirements.yml'
  - '!roles/requirements.yml'

rules:
  braces:
    max-spaces-inside: 1
    level: error
  brackets:
    max-spaces-inside: 1
    level: error
  line-length: disable
  indentation:
    spaces: consistent
    indent-sequences: consistent
```

Important note: the `.yamllint` compatibility warning was present in both the previously passing run and the failing run, so this does not appear to be solely caused by the `.yamllint` file.

Previously passing run:

```console
$ pip3 install ansible-lint
Successfully installed ansible-compat-26.3.0 ansible-core-2.19.11 ansible-lint-26.4.0 yamllint-1.38.0 ...

$ ansible-lint --version
ansible-lint 26.4.0 using ansible-core:2.19.11 ansible-compat:26.3.0 ruamel-yaml:0.19.1 ruamel-yaml-clib:0.2.15

$ ansible-lint --offline --profile=basic -p .
WARNING  Skipped installing old role dependencies due to running in offline mode.
WARNING  Skipped installing collection dependencies due to running in offline mode.
WARNING  Found incompatible custom yamllint configuration (.yamllint), please either remove the file or edit it to comply with:
  - comments.min-spaces-from-content must be 1
  - comments-indentation must be false
  - octal-values.forbid-implicit-octal must be true
  - octal-values.forbid-explicit-octal must be true.

Passed: 0 failure(s), 0 warning(s) in 4 files processed of 5 encountered. Profile 'basic' was required, but 'production' profile passed.
```

Failing run:

```console
$ pip3 install ansible-lint
Successfully installed ansible-compat-26.6.0 ansible-core-2.19.11 ansible-lint-26.6.0 yamllint-1.38.0 ...

$ ansible-lint --version
ansible-lint 26.6.0 using ansible-core:2.19.11 ansible-compat:26.6.0 ruamel-yaml:0.19.1 ruamel-yaml-clib:0.2.15

$ ansible-lint --offline --profile=basic -p .
WARNING  Skipped installing old role dependencies due to running in offline mode.
WARNING  Skipped installing collection dependencies due to running in offline mode.
WARNING  Found incompatible custom yamllint configuration (.yamllint), please either remove the file or edit it to comply with:
  - comments.min-spaces-from-content must be 1
  - comments-indentation must be false
  - octal-values.forbid-implicit-octal must be true
  - octal-values.forbid-explicit-octal must be true.

WARNING  Listing 242 violation(s) that are fatal
```

The failing run then reports many historical violations across the repository, including:

```console
yaml[comments]
schema[requirements]
syntax-check[unknown-module]
syntax-check[specific]
yaml[truthy]
yaml[indentation]
yaml[document-start]
yaml[key-duplicates]
name[play]
name[missing]
```

##### Desired Behavior

A minor version upgrade from `26.4.0` to `26.6.0` should not silently change lint target discovery or fatal behavior enough to break an existing CI job that uses the same command line.

If this broader discovery/failure behavior is intentional, it would be helpful for the release notes or documentation to identify the breaking change and provide migration guidance.

Specifically, we would like to know:

1. Was the broader file discovery between `26.4.0` and `26.6.0` intentional?
2. Should repositories now avoid using `.` as the lint target and instead pass explicit lint targets?
3. Is this related to role-root detection, pathspec changes, `.ansible-lint-ignore` handling, custom `.yamllint` handling, or configuration error visibility?
4. What is the recommended migration path for repositories that contain legacy playbooks, generated content, unfinished playbooks, or content that is intentionally excluded?

Possible security bugs should be reported via email to `security@ansible.com`

##### Actual Behavior

With `ansible-lint==26.4.0`, the pipeline passed:

```console
Passed: 0 failure(s), 0 warning(s) in 4 files processed of 5 encountered. Profile 'basic' was required, but 'production' profile passed.
```

With `ansible-lint==26.6.0`, the same CI command failed:

```console
WARNING  Listing 242 violation(s) that are fatal
```

Examples of the newly fatal findings include:

```console
ExecutionEnvironments/unix-kgb/execution-environment.yml:8: yaml[comments][/]: Missing starting space in comment
ExecutionEnvironments/unix-kgb/requirements.yml:1: schema[requirements][/]: $.collections None is not of type 'array'
add_hosts_to_inventory.yml:68:5: syntax-check[unknown-module]: couldn't resolve module/action 'ansible.controller.inventory_source_update'
azure-vm-create.yml:39: yaml[truthy][/]: Truthy value should be one of [false, true]
azure-vm-customize-packer.yml:2: yaml[indentation][/]: Wrong indentation: expected 0 but found 2
azure_gallery_image_info.yml:1: yaml[document-start]: Missing document start "---"
kit_tasks/deploy_cdis_rsyslog.yml:20: yaml[key-duplicates]: Duplication of key "register" in mapping
```

Current workaround:

```console
pip3 install "ansible-lint==26.4.0" "ansible-compat==26.3.0" "yamllint==1.38.0"
ansible-lint --offline --profile=basic -p .
```

Pinning back to `26.4.0` restores the previous behavior.

The key issue is not that the repository has no lint findings. It clearly has historical lint issues. The issue is that upgrading from `26.4.0` to `26.6.0` appears to have changed discovery/fatal behavior enough to turn a passing pipeline into a failing one without a clear migration path.

**Comment by sathyapramod** [comment]
### Root cause

This looks driven by the directory-expansion fix in **26.6.0** via [#5080](https://github.com/ansible/ansible-lint/pull/5080) (for [#5079](https://github.com/ansible/ansible-lint/issues/5079)).

When CI runs `ansible-lint ... .`, `expand_dirs_in_lintables()` used to filter with `filename.startswith(".")`. Relative paths like `playbook.yml` don’t start with `.`, so most files were skipped — matching the old **4 files processed** result.

26.6.0 switched to `path_is_inside(...)`, so `ansible-lint .` correctly includes the full tree. The **242 fatals** appear to be existing issues that 26.4 never saw, not a sudden rule explosion.

**Net:** behavior correction with breaking CI impact for anyone who relied on the old incomplete discovery.

**Workarounds:** pin `ansible-lint==26.4.0`, or stay on 26.6+ and use `exclude_paths` / explicit lint targets / clean up the newly visible violations.

@cidrblock — does this root-cause read match what you expect from [#5080](https://github.com/ansible/ansible-lint/pull/5080)? And should we call this out as a breaking behavior change in release notes / migration guidance?
